File: examples/go/10_pubsub.go
package main

import (
	"context"
	"fmt"
	"time"
	"github.com/redis/go-redis/v9"
)

func main() {
	ctx := context.Background()
	rdb := redis.NewClient(&redis.Options{Addr: "localhost:6379"})

	fmt.Println("--- 10 Pub/Sub ---")

	pubsub := rdb.Subscribe(ctx, "go_channel")
	defer pubsub.Close()

	// Subscriber goroutine
	go func() {
		ch := pubsub.Channel()
		for msg := range ch {
			fmt.Printf("[Sub] Received: %s\n", msg.Payload)
			if msg.Payload == "quit" {
				return
			}
		}
	}()

	time.Sleep(100 * time.Millisecond)

	// Publish
	rdb.Publish(ctx, "go_channel", "Hello from Go")
	time.Sleep(100 * time.Millisecond)
	rdb.Publish(ctx, "go_channel", "quit")

	time.Sleep(500 * time.Millisecond)
}

File: examples/php/10_pubsub.php
<?php
require 'vendor/autoload.php';

echo "--- 10 Pub/Sub ---\n";

// NOTE: PHP scripts are usually synchronous. 
// Standard pattern is running a separate script for subscriber loop.

$pid = pcntl_fork();

if ($pid == -1) {
    die("Could not fork");
} else if ($pid) {
    // Parent - Publisher
    sleep(1); // Wait for sub
    $pub = new Predis\Client();
    echo "[Pub] Sending...\n";
    $pub->publish("php_channel", "Hello from PHP");
    sleep(1);
    $pub->publish("php_channel", "QUIT");
    pcntl_wait($status); // Wait for child
} else {
    // Child - Subscriber
    $sub = new Predis\Client(['read_write_timeout' => -1]);
    $pubsub = $sub->pubSubLoop();
    $pubsub->subscribe("php_channel");

    foreach ($pubsub as $message) {
        if ($message->kind === 'message') {
            echo "[Sub] Received: {$message->payload}\n";
            if ($message->payload === 'QUIT') {
                $pubsub->unsubscribe();
                break;
            }
        }
    }
}
?>
